_posts/2018-12-06-pmreorder-basics.md

```diff
@@ -0,0 +1,318 @@
+title: Pmreorder basics
+author: wlemkows
+layout: post
+identifier: pmreorder-basics
+---
+
+
+### Introduction
+
+
+When you write your persistent memory application the good practice is
+to check it under pmemcheck - the tool which we have already described in two
+blog posts some time ago. Feel free to come back to them and remind details
+about [basics usage] [pmemcheck1Link] and [transaction][pmemcheck2Link].
+
+In this post, I am going to introduce another tool for persistence correctness check.
+As you already know, pmemcheck verify if all stores are made persistent in a proper
+manner, pmreorder does not replace this functionality but provide extended one.
+Pmreorder traverse sequences of stores between flush-fence barriers made by the application.
+As the next step tool repeats those memory operations many times in different combinations
+to simulate possible order of stores on the NVDIMM.
+
+Each combination of stores is then verified by user-defined consistency checker and
+any inconsistencies are logged. This process allows detecting not persistent data
+on system failure even if pmemcheck does't detect any issue.
+
+It may seem complicated so I will try to explain it using a simple example.
+
+
+### Example
+
+Let's assume that we have a simple list of three elements.
+Code of the list insert function can look like this:
+
+{% highlight C linenos %}
+static void
+list_insert(struct list_root *root, node_id node, int value)
+{
+	struct list_node *new = NODE_PTR(root, node);
+
+	new->next = root->head;
+	pmem_persist(&new->next, sizeof(node));
+
+	root->head = node;
+	pmem_persist(&root->head, sizeof(root->head));
+
+	new->value = value;
+	pmem_persist(&new->value, sizeof(value));
+}
+{% endhighlight %}
+
+Full implementation and explanation you can find in [pmdk repository][pmreorderExampleLink].
+
+At first, the next value in the node is set and persisted.
+Then node is linked at the beginning of the list, its value is set
+and both persisted separately.
+
+When you run this code under pmemcheck you will get the following result:
+ 
+{% highlight sh %}
+==22161== Number of stores nod made persistent: 0
+==22161== ERROR SUMMARY: 0 errors
+{% endhighlight %}
+
+It means that all stores are persisted properly, as we expected.
+
+Now it is time to check the list under pmreorder tool.
+To do that you need to consider when your list is consistent
+and on that basics write consistency checker function.
+
+Follow the next chapters to get to know more details about that.
+
+
+### Checker
+
+Consistency checker is just a binary/function that defines
+conditions necessary to fulfill your consistency assumptions.
+It should return 0 if the state is consistent or 1 if it isn't.
+
+In case of our list with three elements:
+
+{% highlight C linenos %}
+ // params: root, node_id, value
+list_insert(r, 5, 55);
+list_insert(r, 3, 33);
+list_insert(r, 6, 66);
+{% endhighlight %}
+
+we would like to be sure that if node (5, 3, 6) is linked to the list
```

Comment:
if nodes are present in the list, then their value is set properly.

But I'm not sure if that's what you meant.